Add certificates.k8s.io/v1alpha1 ClusterTrustBundle

ahmedtd opened the pull request:
This commit is the main API piece of KEP-3257 (ClusterTrustBundles).

This commit:

* Adds the certificates.k8s.io/v1alpha1 API group
* Adds the ClusterTrustBundle type.
* Registers the new type in kube-apiserver.
* Implements the type-specfic validation specified for ClusterTrustBundles:
  - spec.pemTrustAnchors must always be non-empty.
  - spec.signerName must be either empty or a valid signer name.
  - Changing spec.signerName is disallowed.
* Implements the "entrust" admission check to restrict actions on ClusterTrustBundles that include a signer name.

Because it wasn't specified in the KEP, I chose to make attempts to update the signer name be validation errors, rather than silently ignored.

Additionally, the KEP mentions that the order of trust anchors within the pemTrustAnchors field has no meaning.  I chose to ensure this by sorting, deduplicating, and trimming the pemTrustAnchors field.

I have tested this out by launching these changes in kind and manipulating ClusterTrustBundle objects in the resulting cluster using kubectl.

/kind feature
/kind api-change
/sig auth

```release-note
Added a new alpha API: ClusterTrustBundle (`certificates.k8s.io/v1alpha1`).
A ClusterTrustBundle may be used to distribute [X.509](https://www.itu.int/rec/T-REC-X.509) trust anchors to workloads within the cluster.
```

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/bddca24910fb349e2eb0ac1c822c77f0f32fe9c6/keps/sig-auth/3257-trust-anchor-sets
```

ahmedtd commented:
/assign mikedanese
/assign enj
/assign deads2k

One open question I have: currently, in strategy.go, I do some normalization of the pemTrustAnchors field during Validate, rather than during PrepareForCreate or Canonicalize.  I can't do it in PrepareForCreate because the normalization can report errors, and I can't do it only in Canonicalize because I need access to the normalized data in Validate. 

Doing the normalization in Validate seems to work, but I'm unsure if there's some contract I'm violating.

ahmedtd commented:
If you do want to test this out in Kind, you can use the following cluster config file to enable serving alpha APIs.

```
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
runtimeConfig:
  "api/alpha": "true"
nodes:
- role: control-plane
- role: worker
```

ahmedtd commented:
/label api-review
/milestone 1.26

mikedanese commented on pkg/apis/certificates/types.go:
```diff
@@ -222,3 +222,54 @@ const (
+	//
+	// In order to create or update a ClusterTrustBundle that sets signerName,
+	// you must have the following permission: group=certificates.k8s.io
+	// resource=signers resourceName=<the signer name> verb=entrust.
```
"entrust" feels like the wrong verb. Are you entrusting the signer by creating or updating its trust bundle? IIUC, the trust feels reversed.

ahmedtd replied:
`entrust` is what's in the KEP, but we could go super literal: `settrustanchors`

ahmedtd replied:
@enj Any objection to `settrustanchors`?

mikedanese replied:
What about something very literal like updateclustertrustbundle?

leilajal commented:
/remove-sig api-machinery

enj commented on pkg/apis/certificates/v1alpha1/register.go:
```diff
@@ -0,0 +1,37 @@
+package v1alpha1
```
Missing copyright on this file.

ahmedtd replied:
Done.

enj commented on pkg/apis/certificates/v1alpha1/conversion.go:
```diff
@@ -0,0 +1,21 @@
+package v1alpha1
```
Missing copyright.

ahmedtd replied:
Done.

enj commented on pkg/apis/certificates/v1alpha1/conversion.go:
```diff
@@ -0,0 +1,21 @@
+		SchemeGroupVersion.WithKind("ClusterTrustBundle"),
+		func(label, value string) (string, string, error) {
+			switch label {
+			case "metadata.name", "spec.signerName":
```
Should have an integration test that asserts that the field selector works.

ahmedtd replied:
Done.

enj commented on plugin/pkg/admission/certificates/entrusting/admission.go:
```diff
@@ -0,0 +1,94 @@
+	"k8s.io/kubernetes/plugin/pkg/admission/certificates"
+)
+
+const PluginName = "ClusterTrustBundleEntrusting"
```
If we don't use the `entrust` verb in the authz check then we should probably change the name of the admission plugin too.

ahmedtd replied:
Done.

enj commented on plugin/pkg/admission/certificates/entrusting/admission.go:
```diff
@@ -0,0 +1,94 @@
+		return admission.NewForbidden(a, fmt.Errorf("expected type ClusterTrustBundle, got: %T", a.GetOldObject()))
+	}
+
+	// Unlike CSRs, it's OK to validate against the *new* object, because
+	// updates to signer name will be rejected during validation.  For defense
+	// in depth, reject attempts to change signer at this layer as well.
+	//
+	// We want to use the new object because we also need to perform the signer
+	// name permission check on *create*.
+
+	if a.GetOperation() == admission.Update {
+		oldBundle, ok := a.GetOldObject().(*api.ClusterTrustBundle)
+		if !ok {
+			return admission.NewForbidden(a, fmt.Errorf("expected type ClusterTrustBundle, got: %T", a.GetOldObject()))
+		}
+
+		if oldBundle.Spec.SignerName != newBundle.Spec.SignerName {
+			return admission.NewForbidden(a, fmt.Errorf("changing signerName is forbidden"))
+		}
+	}
```
Just write an integration test to prove this is not possible?

ahmedtd replied:
Done.

enj reviewed (commented):
I still need to look at the unit tests, but I would expect to see integration tests to assert the high level invariants / features of this API.

ahmedtd committed:
ClusterTrustBundles: Define types

This commit is the main API piece of KEP-3257 (ClusterTrustBundles).

This commit:

* Adds the certificates.k8s.io/v1alpha1 API group
* Adds the ClusterTrustBundle type.
* Registers the new type in kube-apiserver.
* Implements the type-specfic validation specified for
  ClusterTrustBundles:
  - spec.pemTrustAnchors must always be non-empty.
  - spec.signerName must be either empty or a valid signer name.
  - Changing spec.signerName is disallowed.
* Implements the "attest" admission check to restrict actions on
  ClusterTrustBundles that include a signer name.

Because it wasn't specified in the KEP, I chose to make attempts to
update the signer name be validation errors, rather than silently
ignored.

I have tested this out by launching these changes in kind and
manipulating ClusterTrustBundle objects in the resulting cluster using
kubectl.

ahmedtd commented on pkg/registry/certificates/rest/storage_certificates.go:
```diff
@@ -60,6 +67,20 @@ func (p RESTStorageProvider) v1Storage(apiResourceConfigSource serverstorage.API
+func (p RESTStorageProvider) v1alpha1Storage(apiResourceConfigSource serverstorage.APIResourceConfigSource, restOptionsGetter generic.RESTOptionsGetter) (map[string]rest.Storage, error) {
+	storage := map[string]rest.Storage{}
+
+	if apiResourceConfigSource.ResourceEnabled(certificatesapiv1alpha1.SchemeGroupVersion.WithResource("clustertrustbundles")) {
```
Does storage of v1alpha1 types need to be behind a feature gate?  Isn't the type being v1alpha1 (which requires explicit runtime configuration passed to kube-apiserver to serve) enough?

I guess in this case it makes sense (since we don't want to server ClusterTrustBundle objects without the fancy admission checks), but the PR I was following (#109090) doesn't do any additional feature gates.  (It also doesn't guard its changes to node authorizer)

enj replied:
Just as we feature gated the alpha `selfsubjectreviews` storage, I would expect us to feature gate this storage as well.

ahmedtd replied:
Done.

ahmedtd commented on pkg/registry/certificates/rest/storage_certificates.go:
```diff
@@ -60,6 +67,20 @@ func (p RESTStorageProvider) v1Storage(apiResourceConfigSource serverstorage.API
+	if apiResourceConfigSource.ResourceEnabled(certificatesapiv1alpha1.SchemeGroupVersion.WithResource("clustertrustbundles")) {
+		bundleStorage, err := clustertrustbundlestore.NewREST(restOptionsGetter)
+		if err != nil {
+			return storage, err
```
If someone registers storage for another type above this block, the code will become silently incorrect.

enj replied:
Any error returned from here will cause the caller to ignore `storage`, and for the server to fail to start.

ahmedtd replied:
Done (here and above)

ahmedtd commented on staging/src/k8s.io/api/certificates/v1alpha1/register.go:
```diff
@@ -0,0 +1,61 @@
+	return SchemeGroupVersion.WithKind(kind).GroupKind()
+}
+
+// Resource takes an unqualified resource and returns a Group qualified GroupResource
+func Resource(resource string) schema.GroupResource {
+	return SchemeGroupVersion.WithResource(resource).GroupResource()
+}
```
Done

ahmedtd replied:
Actually, these appear to be needed from the generated clients.

ahmedtd commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -150,6 +150,9 @@ func NodeRules() []rbacv1.PolicyRule {
 		rbacv1helpers.NewRule("create", "get", "list", "watch").Groups(certificatesGroup).Resources("certificatesigningrequests").RuleOrDie(),
 
+		// Needed for ClusterTrustBundle / pemTrustAnchors projected volumes.
+		rbacv1helpers.NewRule("get", "list", "watch").Groups(certificatesGroup).Resources("clustertrustbundles").RuleOrDie(),
```
Done.  (Documentation added on the ClusterTrustBundle type).

ahmedtd commented on staging/src/k8s.io/api/certificates/v1alpha1/types.go:
```diff
@@ -0,0 +1,77 @@
+// contains one valid set of trust anchors for that signer. Signers may have
+// multiple associated ClusterTrustBundles; each is an independent set of trust
+// anchors for that signer. Admission control is used to enforce that only
+// users with permissions on the signer can create or modify the corresponding.
```
Done.

ahmedtd commented on staging/src/k8s.io/api/certificates/v1alpha1/types.go:
```diff
@@ -0,0 +1,77 @@
+// users with permissions on the signer can create or modify the corresponding.
+type ClusterTrustBundle struct {
+	metav1.TypeMeta `json:",inline"`
+	//+optional
```
Done.

ahmedtd commented on staging/src/k8s.io/api/certificates/v1alpha1/types.go:
```diff
@@ -0,0 +1,77 @@
+	// pemTrustAnchors contains the individual X.509 trust anchors for this
+	// bundle, as PEM bundle of PEM-wrapped, DER-formatted X.509 certificates.
+	//
+	// The order of certificates within the bundle has no meaning.
```
As-is the validation will strip them out (as well as PEM headers in each block).

I think when we wrote that we would preseve comments in the KEP, we didn't think through the implications.  Specifically, to *force* that the order is unimportant, we should stably order the certificates according to some criteria (and perhaps de- and re-serialize the certificates to remove differences in PEM formatting).  Doing that makes preserving the inter-block comments much harder.

liggitt replied:
> Specifically, to force that the order is unimportant, we should stably order the certificates according to some criteria (and perhaps de- and re-serialize the certificates to remove differences in PEM formatting). Doing that makes preserving the inter-block comments much harder.

just picking up the review, but this caught my eye... normalizing wreaks havoc on server-side apply which infers differences between the persisted object and the incoming object in an update :-/

enj replied:
https://github.com/kubernetes/kubernetes/pull/113218#discussion_r1014537571

ahmedtd replied:
OK, now we just do validation on the API server.  The Kubelet half will do any normalization.

ahmedtd replied:
> normalizing wreaks havoc on server-side apply which infers differences between the persisted object and the incoming object in an update

Probably this should be made obvious from the code / forbidden?  It sounds like no new types should be using this capability.

ahmedtd commented on pkg/registry/certificates/clustertrustbundle/strategy.go:
```diff
@@ -0,0 +1,147 @@
+
+func (strategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {}
+
+func noRestrictionsOnName(name string, prefix bool) []string {
+	return nil
+}
```
I don't think so, because some ClusterTrustBundles might not be associated with a signer (like in the webhook configuration use case).

enj replied:
Right but it may be a good idea to encourage people to call things `gatekeeper-io--webhook` vs `webhook`.

ahmedtd replied:
Yeah, I thought about it a little more and I think you're right.  I will add this.

ahmedtd replied:
Done.  Now if spec.signerName is set, we enforce a naming prefix.

ahmedtd replied:
I think we touched on this during API review, and explored whether or not we should be enforcing a name format, but we failed to come up with a simple escaping scheme.  We decided that if Kubelet could mount ClusterTrustBundles directly by signer name, then the prefix was less important.

I've changed the projected volume PR to allow mounting 1) either a single CTB by name, or 2) A group of CTBs by signer name and label selector.  I think this obviates the need to define a strict naming scheme, so I will drop this piece of the validation.  We will still reserve the `kubernetes-io--` prefix for future first-party use.

ahmedtd commented on pkg/registry/certificates/clustertrustbundle/strategy.go:
```diff
@@ -0,0 +1,147 @@
+		allErrors = append(allErrors, signerNameErrors...)
+	}
+
+	// TODO(KEP-3257): Is it OK to modify the object during validate?
```
What do you think we should do here, then?

* PrepareForCreate can't report an error, and the pemTrustAnchors normalization step can encounter lots of errors.
* If we deep-copy the object passed to Validate() so that we don't modify the input object, then we need to re-run the sorting and de-re-serializing of certificates twice (once in PrepareForCreate or Canonicalize, and once in Validate).
* Canonicalize() also can't report an error, so we would need to trust that there's no modification of the object between Validate() and Canonicalize() that would cause pemTrustAnchors normalization to fail.

If we're going to trust what the code calling our strategy does, we might as well trust what it does today, which has no problem with a mutation during Validate() (though we would have to update all the comments that say that Validate() should not mutate the object): https://github.com/kubernetes/kubernetes/blob/66918763265b91000482f82a7e92c68d24733542/staging/src/k8s.io/apiserver/pkg/registry/rest/create.go#L120

In general, validation and canonicalization can't be cleanly separated without duplicating work.  The current code structure seems a little bit optimistic that they can.

ahmedtd replied:
I decided to just do the duplication for now.  We can always optimize later.

enj replied:
https://github.com/kubernetes/kubernetes/pull/113218#discussion_r1014537571

ahmedtd replied:
OK, now we just do validation on the API server. The Kubelet half will do any normalization.

ahmedtd replied:
I think this conversation is resolved.  Validation no longer changes the user's input, just rejects it.

ahmedtd commented on pkg/registry/certificates/clustertrustbundle/strategy.go:
```diff
@@ -0,0 +1,147 @@
+	return nil
+}
+
+func (strategy) AllowUnconditionalUpdate() bool {
+	return true
+}
```
Copied from #109090.

You're right, probably safer to forbid it.

ahmedtd commented on pkg/registry/certificates/clustertrustbundle/strategy.go:
```diff
@@ -0,0 +1,147 @@
+	return true
+}
+
+// normalizePEMTrustAnchors strips interblock data, strips in-block headers,
+// reserializes the blocks to remove line-wrapping and padding differences, then
+// sorts them alphabetically.
```
Basically to prevent clients from depending on having a specific order available.  It would be even better if we could find a way to make the order stable but unpredictable (maybe by keying the sort order on the overall hash of all the certificates)?

In-block headers are cut because we re-encode the PEM blocks to remove linewrapping differences (and also because in-block headers are a bad idea).

Interblock comments are cut as a consequence of sorting the certs (and also because they're not a great idea, people will be tempted to use them for signaling).

enj replied:
I do not think we should mutate the user's input in any way.  If the input is valid, it should be persisted as-is.  It makes the API contract nicer for humans.

However, as far as code consuming this is concerned:

Randomizing order is good.
Normalizing line wrapping is good.
Removing comments is good.

All of that should be done by kubelet before it mounts stuff into pods.  If direct consumers of the REST API want to be bad, they could simply use labels and annotations to do so.

ahmedtd replied:
Hmmm, I didn't think about having Kubelet do this.  You're right, that's a better idea.

ahmedtd replied:
I think this is done.  We no longer mutate user input during validation.

ahmedtd commented on pkg/registry/certificates/clustertrustbundle/strategy.go:
```diff
@@ -0,0 +1,147 @@
+// reserializes the blocks to remove line-wrapping and padding differences, then
+// sorts them alphabetically.
+func normalizePEMTrustAnchors(in string) (string, error) {
+	blockSet := map[string]bool{}
```
Done.

ahmedtd replied:
Actually, undone since I changed this around to be validate-only.  Now we check that we didn't get multiple copies of the same trust anchor.

ahmedtd commented on pkg/registry/certificates/clustertrustbundle/strategy.go:
```diff
@@ -0,0 +1,147 @@
+func normalizePEMTrustAnchors(in string) (string, error) {
+	blockSet := map[string]bool{}
+
+	// TODO(KEP-3257): Discuss how protective to be of downstream systems.
+	// Should we verify that the PEM data parses as an X.509 certificate?
```
Done.

ahmedtd commented on pkg/registry/certificates/clustertrustbundle/strategy.go:
```diff
@@ -0,0 +1,147 @@
+	return false
+}
+
+func (s strategy) PrepareForUpdate(ctx context.Context, new, old runtime.Object) {
+	s.PrepareForCreate(ctx, new)
+}
```
I was worried that I would come back and add normalization to PrepareForCreate, but forget to duplicate it into PrepareForUpdate.  Now that the normalization lives in Canonicalize(), we can remove this.

ahmedtd replied:
Done.

ahmedtd commented on pkg/registry/certificates/clustertrustbundle/strategy.go:
```diff
@@ -0,0 +1,147 @@
+	return nil
+}
+
+func (strategy) Validate(ctx context.Context, obj runtime.Object) field.ErrorList {
```
Done.

ahmedtd commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -150,6 +150,9 @@ func NodeRules() []rbacv1.PolicyRule {
 		// for it to be signed. This allows the kubelet to rotate it's own certificate.
 		rbacv1helpers.NewRule("create", "get", "list", "watch").Groups(certificatesGroup).Resources("certificatesigningrequests").RuleOrDie(),
 
+		// Needed for ClusterTrustBundle / pemTrustAnchors projected volumes.
+		rbacv1helpers.NewRule("get", "list", "watch").Groups(certificatesGroup).Resources("clustertrustbundles").RuleOrDie(),
+
```
Done.

ahmedtd commented on api/openapi-spec/swagger.json:
```diff
@@ -3549,6 +3549,88 @@
       },
       "type": "object"
     },
+    "io.k8s.api.certificates.v1alpha1.ClusterTrustBundle": {
```
Done.

ahmedtd commented on plugin/pkg/admission/certificates/entrusting/admission.go:
```diff
@@ -0,0 +1,94 @@
+
+var clusterTrustBundleGroupResource = api.Resource("clustertrustbundles")
+
+func (p *Plugin) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) error {
```
Done.

ahmedtd commented on pkg/registry/certificates/clustertrustbundle/storage/storage.go:
```diff
@@ -0,0 +1,67 @@
+	return &REST{store}, nil
+}
+
+var _ rest.ShortNamesProvider = &REST{}
```
Done (I think).

ahmedtd commented on pkg/registry/certificates/clustertrustbundle/storage/storage.go:
```diff
@@ -0,0 +1,67 @@
+
+var _ rest.ShortNamesProvider = &REST{}
+
+// ShortNames returns a list of short names for ClusterTrustBundle.
+//
+// Implements the ShortNamesProvider interface.
+func (r *REST) ShortNames() []string {
+	return []string{"ctb"}
+}
```
Removed.

ahmedtd commented on staging/src/k8s.io/api/certificates/v1alpha1/types.go:
```diff
@@ -0,0 +1,77 @@
+// ClusterTrustBundle is a cluster-scoped container for X.509 trust anchors
+// (root certificates).
+//
+// ClusterTrustBundle objects are considered to be readable by any authenticated
+// user in the cluster.
```
Done.

ahmedtd commented on staging/src/k8s.io/api/certificates/v1alpha1/types.go:
```diff
@@ -0,0 +1,77 @@
+	// signerName indicates the associated signer, if any.
+	//
+	// In order to create or update a ClusterTrustBundle that sets signerName,
+	// you must have the following permission: group=certificates.k8s.io
+	// resource=signers resourceName=<the signer name> verb=entrust.
```
Done.

ahmedtd committed:
ClusterTrustBundles: make update

ahmedtd commented on staging/src/k8s.io/api/certificates/v1alpha1/register.go:
```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Done

ahmedtd commented:
OK, I've uploaded a new version with all existing feedback addressed (I think).

* Regeneration is now in a separate commit.
* Storage and admission is now feature-gated.
* We now only (aggressively) validate ClusterTrustBundles.  Normalization is left for Kubelet.
* The verb is renamed to "establishtrust".
* ClusterTrustBundles with signer name set to `foo.com/bar` are now required to have the name prefix `foo-com--bar`.

Still need to add integration tests.

ahmedtd commented:
/retest

ahmedtd committed:
ClusterTrustBundles: kube-apiserver integration tests

ahmedtd commented:
I've added integration tests, which I think addresses the last open comment

pacoxu commented on test/integration/clustertrustbundles/signer_name_change_forbidden_test.go:
```diff
@@ -0,0 +1,102 @@
+			})),
+		},
+	}
+	bundle1, err := client.CertificatesV1alpha1().ClusterTrustBundles().Create(ctx, bundle1, metav1.CreateOptions{})
```
```suggestion
	_, err := client.CertificatesV1alpha1().ClusterTrustBundles().Create(ctx, bundle1, metav1.CreateOptions{})
```

ahmedtd replied:
Done.

pacoxu commented on pkg/apis/certificates/v1alpha1/conversion.go:
```diff
@@ -0,0 +1,21 @@
+package v1alpha1
```
```suggestion
/*
Copyright 2022 The Kubernetes Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package v1alpha1
```

ahmedtd replied:
Done.

enj commented:
/milestone v1.27

We decided to defer this feature until next release.

dims commented:
If you still need this PR then please rebase, if not, please close the PR

liggitt commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -537,3 +537,118 @@ func hasDuplicateUsage(usages []certificates.KeyUsage) bool {
+			continue
+		}
+
+		cert, err := x509.ParseCertificate(b.Bytes)
```
consider how to insulate ourselves from changes (relaxing or tightening) to the golang stdlib ParseCertificate method... at minimum, on update, if the old and new object have identical PEMTrustAnchors, the update should not fail validation (will require a little plumbing from the Validate___Update call... there's a `ValidationOptions { ValidateCertificateContent bool }` pattern you could pass down to the point where we validate this field)

liggitt replied:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-enum-value-in-existing-field has an example of the validation options struct pattern

ahmedtd replied:
Done.

liggitt commented on pkg/registry/certificates/clustertrustbundle/strategy.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+func (strategy) AllowUnconditionalUpdate() bool {
+	return true
```
change this to false

ahmedtd replied:
Done.

liggitt commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -537,3 +537,118 @@ func hasDuplicateUsage(usages []certificates.KeyUsage) bool {
+	allErrors = append(allErrors, ValidateClusterTrustBundle(newBundle)...)
+	allErrors = append(allErrors, apivalidation.ValidateObjectMetaUpdate(&newBundle.ObjectMeta, &oldBundle.ObjectMeta, field.NewPath("metadata"))...)
+
+	if newBundle.Spec.SignerName != oldBundle.Spec.SignerName {
```
use ValidateImmutableField() for this, and can drop the check from admission

ahmedtd replied:
Done.

liggitt commented on pkg/registry/certificates/rest/storage_certificates.go:
```diff
@@ -60,6 +70,24 @@ func (p RESTStorageProvider) v1Storage(apiResourceConfigSource serverstorage.API
+func (p RESTStorageProvider) v1alpha1Storage(apiResourceConfigSource serverstorage.APIResourceConfigSource, restOptionsGetter generic.RESTOptionsGetter) (map[string]rest.Storage, error) {
+	storage := map[string]rest.Storage{}
+
+	if apiResourceConfigSource.ResourceEnabled(certificatesapiv1alpha1.SchemeGroupVersion.WithResource("clustertrustbundles")) {
```
use the:

```
if resource := "clustertrustbundles"; ...
  storage[resource] = ...
```

pattern to avoid setting up copy/paste landmines

ahmedtd replied:
Done.

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -609,6 +625,11 @@ func ClusterRoleBindings() []rbacv1.ClusterRoleBinding {
 
+	// Service accounts can read ClusterTrustBundle objects.
+	if utilfeature.DefaultFeatureGate.Enabled(features.ClusterTrustBundle) {
+		rolebindings = append(rolebindings, rbacv1helpers.NewClusterBinding("system:cluster-trust-bundle-discovery").Groups(serviceaccount.AllServiceAccountsGroup).BindingOrDie())
```
I think this is reasonable, but make a note in the KEP to double-check this before graduation to beta

liggitt reviewed (commented):
add the new group/version to staging/src/k8s.io/api/roundtrip_test.go to exercise the new types in round-trip testing

liggitt commented:
update the etcd test to enable feature gates to catch any gated-APIs:

```diff
diff --git a/test/integration/etcd/etcd_storage_path_test.go b/test/integration/etcd/etcd_storage_path_test.go
index 5014506169e..c5ecbe42816 100644
--- a/test/integration/etcd/etcd_storage_path_test.go
+++ b/test/integration/etcd/etcd_storage_path_test.go
@@ -35,7 +35,9 @@ import (
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/util/diff"
 	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/util/feature"
 	"k8s.io/client-go/dynamic"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
 	"k8s.io/kubernetes/cmd/kube-apiserver/app/options"
 )
 
@@ -70,6 +72,8 @@ var allowMissingTestdataFixtures = map[schema.GroupVersionKind]bool{
 // It will also fail when a type gets moved to a different location. Be very careful in this situation because
 // it essentially means that you will be break old clusters unless you create some migration path for the old data.
 func TestEtcdStoragePath(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, "AllAlpha", true)()
+	defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, "AllBeta", true)()
 	apiServer := StartRealAPIServerOrDie(t, func(opts *options.ServerRunOptions) {
 	})
 	defer apiServer.Cleanup()
```

liggitt commented on plugin/pkg/admission/certificates/establishtrust/admission.go:
```diff
@@ -0,0 +1,118 @@
+		return nil
+	}
+
+	if !certificates.IsAuthorizedForSignerName(ctx, p.authz, a.GetUserInfo(), "establishtrust", newBundle.Spec.SignerName) {
```
modeling this as admission now is fine, make a note in the KEP to confirm this decision before beta in case the fine-grained authz KEP has changed our thinking on doing this in storage

enj commented:
> update the etcd test to enable feature gates to catch any gated-APIs:

I have opened #114885 for this.

ahmedtd commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -537,3 +537,118 @@ func hasDuplicateUsage(usages []certificates.KeyUsage) bool {
+	allErrors = append(allErrors, ValidateClusterTrustBundle(newBundle)...)
+	allErrors = append(allErrors, apivalidation.ValidateObjectMetaUpdate(&newBundle.ObjectMeta, &oldBundle.ObjectMeta, field.NewPath("metadata"))...)
+
+	if newBundle.Spec.SignerName != oldBundle.Spec.SignerName {
+		allErrors = append(allErrors, field.Forbidden(field.NewPath("spec", "signerName"), "updates may not change the signer name"))
+	}
```
Done.

ahmedtd commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -537,3 +537,118 @@ func hasDuplicateUsage(usages []certificates.KeyUsage) bool {
+// validatePEMTrustAnchors rejects intra-block headers, blocks
+// that don't parse as X.509 CA certificates, and duplicate trust anchors.  It
+// requires that at least one trust anchor is provided.
+func validatePEMTrustAnchors(in string) field.ErrorList {
```
Done.

ahmedtd commented on pkg/apis/certificates/types.go:
```diff
@@ -222,3 +222,54 @@ const (
+	//
+	// In order to create or update a ClusterTrustBundle that sets signerName,
+	// you must have the following permission: group=certificates.k8s.io
+	// resource=signers resourceName=<the signer name> verb=establishtrust.
```
Done.

ahmedtd commented on pkg/apis/certificates/types.go:
```diff
@@ -222,3 +222,54 @@ const (
+// It can be optionally associated with a particular assigner, in which case it
+// contains one valid set of trust anchors for that signer. Signers may have
+// multiple associated ClusterTrustBundles; each is an independent set of trust
+// anchors for that signer. Admission control is used to enforce that only
+// users with permissions on the signer can create or modify the corresponding.
```
Done.

ahmedtd commented on pkg/apis/certificates/types.go:
```diff
@@ -222,3 +222,54 @@ const (
+	// PEMTrustAnchors contains the individual X.509 trust anchors for this
+	// bundle, as PEM bundle of PEM-wrapped, DER-formatted X.509 certificates.
+	//
+	// The order of certificates within the bundle has no meaning.
```
Done.

ahmedtd commented on staging/src/k8s.io/api/certificates/v1alpha1/types.go:
```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
Done.

ahmedtd commented on staging/src/k8s.io/api/certificates/v1alpha1/types.go:
```diff
@@ -0,0 +1,86 @@
+	// resource=signers resourceName=<the signer name> verb=establishtrust.
+	//
+	// +optional
+	SignerName string `json:"signerName" protobuf:"bytes,1,opt,name=signerName"`
```
Done

ahmedtd commented on staging/src/k8s.io/api/certificates/v1alpha1/types.go:
```diff
@@ -0,0 +1,86 @@
+	// In order to create or update a ClusterTrustBundle that sets signerName,
+	// you must have the following cluster-scoped permission: group=certificates.k8s.io
+	// resource=signers resourceName=<the signer name> verb=establishtrust.
+	//
```
I ended up dropping the name restrictions and instead giving the Kubelet projected volume the ability to mount bundles by signer name and label selector combo.

ahmedtd replied:
Separately, should this file contain mention of things that happen via admission control?

liggitt replied:
validation looks good, update the docs here with the metadata.name requirements for objects that set signerName or leave signerName empty

ahmedtd replied:
Done.

ahmedtd replied:
Oops, I see a typo.  Will fix.

ahmedtd commented on pkg/registry/certificates/clustertrustbundle/strategy.go:
```diff
@@ -0,0 +1,81 @@
+func (s strategy) ValidateUpdate(ctx context.Context, new, old runtime.Object) field.ErrorList {
+	oldBundle := old.(*certificates.ClusterTrustBundle)
+	newBundle := new.(*certificates.ClusterTrustBundle)
+	return certvalidation.ValidateClusterTrustBundleUpdate(oldBundle, newBundle)
```
Done.

ahmedtd commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -537,3 +537,118 @@ func hasDuplicateUsage(usages []certificates.KeyUsage) bool {
+
+// ValidateClusterTrustBundleUpdate runs all update validation checks on an
+// update.
+func ValidateClusterTrustBundleUpdate(oldBundle, newBundle *certificates.ClusterTrustBundle) field.ErrorList {
```
Done.

ahmedtd commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -537,3 +537,118 @@ func hasDuplicateUsage(usages []certificates.KeyUsage) bool {
+		if signerName != "" {
+			escaped := strings.ReplaceAll(signerName, ".", "-")
+			escaped = strings.ReplaceAll(escaped, "/", "--")
+			if !strings.HasPrefix(name, escaped) {
```
I decided to drop name restrictions.  Instead, users of the Kubelet projected volume will be able to select bundles to mount by a combination of signerName and label selector.  This allows the names of the ClusterTrustBundle objects to be arbitrary.

ahmedtd commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -537,3 +537,118 @@ func hasDuplicateUsage(usages []certificates.KeyUsage) bool {
+func requireSignerNameAsPrefix(signerName string) func(name string, prefix bool) []string {
+	return func(name string, prefix bool) []string {
+		if signerName != "" {
+			escaped := strings.ReplaceAll(signerName, ".", "-")
```
Ditto above --- I decided to just drop name restrictions.

ahmedtd commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -537,3 +537,118 @@ func hasDuplicateUsage(usages []certificates.KeyUsage) bool {
+	return func(name string, prefix bool) []string {
+		if signerName != "" {
+			escaped := strings.ReplaceAll(signerName, ".", "-")
+			escaped = strings.ReplaceAll(escaped, "/", "--")
```
Ditto above --- I decided to just drop name restrictions.

ahmedtd commented:
This is updated, and all outstanding review comments are addressed (except for two which request recording some open questions in the KEP).

liggitt commented on plugin/pkg/admission/certificates/establishtrust/admission.go:
```diff
@@ -0,0 +1,118 @@
+		return nil
+	}
+
+	if !certificates.IsAuthorizedForSignerName(ctx, p.authz, a.GetUserInfo(), "establishtrust", newBundle.Spec.SignerName) {
```
I was talking with @enj to see if there was a less awkward verb we could use here... what would you think of `attest`? "affirm to be true or genuine", which is what trust bundles do for certificates, so anyone with write access is doing that for stuff issued by that signer.

Has the benefit of being an actual verb, and being used in security circles

ahmedtd replied:
@mikedanese ?  (I have no real opinion on this verb).

I'll go ahead and assume that Mike is also fine with it, and make the change.

ahmedtd replied:
Done.

liggitt commented:
Looks really close. The name validation and the verb used by admission are the only two functional tweaks remaining.

liggitt commented on staging/src/k8s.io/api/certificates/v1alpha1/types.go:
```diff
@@ -0,0 +1,86 @@
+	// bundle, as PEM bundle of PEM-wrapped, DER-formatted X.509 certificates.
+	//
+	// The order of certificates within the bundle has no meaning.
+	PEMTrustAnchors string `json:"pemTrustAnchors" protobuf:"bytes,2,opt,name=pemTrustAnchors"`
```
oh, I meant to ask if the field name should be consistent with the type `ClusterTrustBundle` → `spec.trustBundle`, and whether we need to include "PEM" in the field name (we don't in any of the other API fields that contain PEM content)

ahmedtd replied:
Done.

ahmedtd commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -537,3 +537,116 @@ func hasDuplicateUsage(usages []certificates.KeyUsage) bool {
 	return false
 }
+
+// We don't care what you call your cluster trust bundles.
+func validateClusterTrustBundleName(name string, prefix bool) []string {
+	return nil
+}
```
Seems reasonable.  I'll do that.

ahmedtd replied:
The standard name validation approach seems to break name generation:

    > resttest.go:375: Unexpected error: ClusterTrustBundle.certificates.k8s.io "test-72l7z" is invalid: metadata.generateName: Invalid value: "test-": a lowercase RFC 1123 subdomain must consist ...

The problem seems to solely be that the common pattern for GenerateName is to end in a dash, but that's not a valid RFC 1123 subdomain.  I could solve this by:

* Passing a no-op name validation to ValidateObjectMeta, and then implement my own checks for Name and GenerateName.
* Extending ValidateObjectMeta to strip a trailing dash from GenerateName before applying validation.
* Something else?

Do you have a suggestion on which way to go?

liggitt replied:
call NameIsDNSSubdomain and pass the `prefix` boolean through, and that function will mask the trailing dash if present

ahmedtd replied:
Done.

ahmedtd commented on test/integration/clustertrustbundles/signer_name_change_forbidden_test.go:
```diff
@@ -0,0 +1,101 @@
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+func TestCTBSignerNameChangeForbidden(t *testing.T) {
```
It's already covered as a unit test; I think I put this as an integration test just because this is security-critical guarantee, so we need to make sure that the validation is actually wired up.

ahmedtd commented on staging/src/k8s.io/api/certificates/v1alpha1/types.go:
```diff
@@ -0,0 +1,86 @@
+	SignerName string `json:"signerName,omitempty" protobuf:"bytes,1,opt,name=signerName"`
+
+	// pemTrustAnchors contains the individual X.509 trust anchors for this
+	// bundle, as PEM bundle of PEM-wrapped, DER-formatted X.509 certificates.
```
Done.

neoaggelos commented:
Hi @ahmedtd bug triage team for 1.27 here. Note that the Code Freeze for 1.27 is in two days March 14th, do you believe this PR will be ready and merged by that time?

ahmedtd commented:
@neoaggelos I believe it should be ready before the freeze.

ahmedtd commented:
@liggitt All suggestions applied.

liggitt commented on staging/src/k8s.io/api/certificates/v1alpha1/types.go:
```diff
@@ -0,0 +1,97 @@
+
+// ClusterTrustBundleSpec contains the signer and trust anchors.
+type ClusterTrustBundleSpec struct {
+	// SignerName indicates the associated signer, if any.
```
```suggestion
	// signerName indicates the associated signer, if any.
```

ahmedtd replied:
Done.

liggitt commented on staging/src/k8s.io/api/certificates/v1alpha1/types.go:
```diff
@@ -0,0 +1,97 @@
+	// +optional
+	SignerName string `json:"signerName,omitempty" protobuf:"bytes,1,opt,name=signerName"`
+
+	// TrustBundle contains the individual X.509 trust anchors for this
```
```suggestion
	// trustBundle contains the individual X.509 trust anchors for this
```

ahmedtd replied:
Done.

liggitt commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -537,3 +538,129 @@ func hasDuplicateUsage(usages []certificates.KeyUsage) bool {
+		if !strings.HasPrefix(name, requiredPrefix) {
+			return []string{fmt.Sprintf("ClusterTrustBundle for signerName %s must be named with prefix %s", signerName, requiredPrefix)}
+		}
+		return apimachineryvalidation.NameIsDNSSubdomain(strings.TrimPrefix(name, requiredPrefix), isPrefix)
```
does this allow or reject an empty string if `name` == `requiredPrefix`? should *probably* reject

ahmedtd replied:
I have a test case (`"invalid, with signer name, empty name suffix"`) covering this; it's rejected.

liggitt commented on test/integration/etcd/data.go:
```diff
@@ -181,6 +181,13 @@ func GetEtcdStorageDataForNamespace(namespace string) map[schema.GroupVersionRes
 
+		// k8s.io/kubernetes/pkg/apis/certificates/v1alpha1
+		gvr("certificates.k8s.io", "v1alpha1", "clustertrustbundles"): {
+			Stub:             `{"metadata": {"name": "ctb1"}, "spec": {"signerName":"example.com/signer", "trustAnchors": "-----BEGIN CERTIFICATE-----\nMIIBBDCBt6ADAgECAgEAMAUGAytlcDAQMQ4wDAYDVQQDEwVyb290MTAiGA8wMDAx\nMDEwMTAwMDAwMFoYDzAwMDEwMTAxMDAwMDAwWjAQMQ4wDAYDVQQDEwVyb290MTAq\nMAUGAytlcAMhAF2MoFeGa97gK2NGT1h6p1/a1GlMXAXbcjI/OShyIobPozIwMDAP\nBgNVHRMBAf8EBTADAQH/MB0GA1UdDgQWBBTWDdK2CNQiHqRjPaAWYPPtIy ...
```
TestEtcdStoragePath caught the name of this needed to be updated to fit the signerName restriction on metadata.name

liggitt replied:
and also caught the field name was invalid, API type used `trustBundle`, not `trustAnchors`

ahmedtd replied:
Done.

liggitt commented on test/integration/clustertrustbundles/field_selector_test.go:
```diff
@@ -0,0 +1,135 @@
+		},
+		Spec: certsv1alpha1.ClusterTrustBundleSpec{
+			SignerName: "baz.com/bar",
+			TrustAnchors: mustMakePEMBlock("CERTIFICATE", nil, mustMakeCertificate(t, &x509.Certificate{
```
ERROR(darwin/arm64): /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/clustertrustbundles/field_selector_test.go:87:4: unknown field TrustAnchors in struct literal of type "k8s.io/kubernetes/vendor/k8s.io/api/certificates/v1alpha1".ClusterTrustBundleSpec

ahmedtd replied:
Done.

liggitt reviewed (commented):
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/113218/pull-kubernetes-verify/1635116627567579136 failed, looks like `make update` is needed (make sure you have latest tags fetched from github when you run that)

liggitt commented on pkg/registry/registrytest/etcd.go:
```diff
@@ -42,11 +42,11 @@ func NewEtcdStorageForResource(t *testing.T, resource schema.GroupResource) (*st
 	resourceConfig, err := factory.NewConfig(resource)
 	if err != nil {
-		t.Fatal(err)
+		t.Fatalf("Error while finding storage destination: %w", err)
```
vet is catching this:
```suggestion
		t.Fatalf("Error while finding storage destination: %v", err)
```

ahmedtd replied:
Done.

ahmedtd replied:
Sorry, I had an adventure with rebasing and then forgot to push this change back down to this PR.  This is done now.

liggitt commented:
doc on the signerName field about the metadata.name requirements is the only substantive comment, then lgtm: https://github.com/kubernetes/kubernetes/pull/113218#discussion_r1134485979

(will need a doc regen and a fix for the vet error 
at https://github.com/kubernetes/kubernetes/pull/113218#discussion_r1134894372)

liggitt commented:
/lgtm
/approve

liggitt commented:
/lgtm

ahmedtd commented:
I'm not sure why verify openapi-spec is failing --- I've run make update

liggitt commented:
looks like a metadata type changed in master, rebasing this PR and generating gives me a diff:

```diff
diff --git a/api/openapi-spec/swagger.json b/api/openapi-spec/swagger.json
index 49e52979cd0..6ff663aa84c 100644
--- a/api/openapi-spec/swagger.json
+++ b/api/openapi-spec/swagger.json
@@ -58426,6 +58426,13 @@
             "type": "string",
             "uniqueItems": true
           },
+          {
+            "description": "`sendInitialEvents=true` may be set together with `watch=true`. In that case, the watch stream will begin with synthetic events to produce the current state of objects in the collection. Once all such events have been sent, a synthetic \"Bookmark\" event  will be sent. The bookmark will report the ResourceVersion (RV) corresponding to the set of objects, and be marked with `\"k8s.io/initial-events-end\": \"true\"` annotation. Afterwards, the watch stream will proceed as usual, sending watch events corresponding to changes (subsequent to the RV) to objects watched.\n\nWhen `sendInitialEvents` option is set, we require `resourceVersionMatch` option to also be set. The semantic of the watch request is as following: - `resourceVersionMatch` = NotOlderThan\n  is interpreted as \"data at least as new as the provided `resourceVersion`\"\n  and the bookmark event is send when the state is synced\n  to a `resourceVersion` at least as fresh as the one provided by the ListOptions.\n  If `resourceVersion` is unset, this is interpreted as \"consistent read\" and the\n  bookmark event is send when the state is synced at least to the moment\n  when request started being processed.\n- `resourceVersionMatch` set to any other value or unset\n  Invalid error is returned.\n\nDefaults to true if `resourceVersion=\"\"` or `resourceVersion=\"0\"` (for backward compatibility reasons) and to false otherwise.",
+            "in": "query",
+            "name": "sendInitialEvents",
+            "type": "boolean",
+            "uniqueItems": true
+          },
           {
             "description": "Timeout for the list/watch call. This limits the duration of the call, regardless of any activity or inactivity.",
             "in": "query",
@@ -58519,6 +58526,13 @@
             "type": "string",
             "uniqueItems": true
           },
+          {
+            "description": "`sendInitialEvents=true` may be set together with `watch=true`. In that case, the watch stream will begin with synthetic events to produce the current state of objects in the collection. Once all such events have been sent, a synthetic \"Bookmark\" event  will be sent. The bookmark will report the ResourceVersion (RV) corresponding to the set of objects, and be marked with `\"k8s.io/initial-events-end\": \"true\"` annotation. Afterwards, the watch stream will proceed as usual, sending watch events corresponding to changes (subsequent to the RV) to objects watched.\n\nWhen `sendInitialEvents` option is set, we require `resourceVersionMatch` option to also be set. The semantic of the watch request is as following: - `resourceVersionMatch` = NotOlderThan\n  is interpreted as \"data at least as new as the provided `resourceVersion`\"\n  and the bookmark event is send when the state is synced\n  to a `resourceVersion` at least as fresh as the one provided by the ListOptions.\n  If `resourceVersion` is unset, this is interpreted as \"consistent read\" and the\n  bookmark event is send when the state is synced at least to the moment\n  when request started being processed.\n- `resourceVersionMatch` set to any other value or unset\n  Invalid error is returned.\n\nDefaults to true if `resourceVersion=\"\"` or `resourceVersion=\"0\"` (for backward compatibility reasons) and to false otherwise.",
+            "in": "query",
+            "name": "sendInitialEvents",
+            "type": "boolean",
+            "uniqueItems": true
+          },
           {
             "description": "Timeout for the list/watch call. This limits the duration of the call, regardless of any activity or inactivity.",
             "in": "query",
@@ -59034,6 +59048,13 @@
           "type": "string",
           "uniqueItems": true
         },
+        {
+          "description": "`sendInitialEvents=true` may be set together with `watch=true`. In that case, the watch stream will begin with synthetic events to produce the current state of objects in the collection. Once all such events have been sent, a synthetic \"Bookmark\" event  will be sent. The bookmark will report the ResourceVersion (RV) corresponding to the set of objects, and be marked with `\"k8s.io/initial-events-end\": \"true\"` annotation. Afterwards, the watch stream will proceed as usual, sending watch events corresponding to changes (subsequent to the RV) to objects watched.\n\nWhen `sendInitialEvents` option is set, we require `resourceVersionMatch` option to also be set. The semantic of the watch request is as following: - `resourceVersionMatch` = NotOlderThan\n  is interpreted as \"data at least as new as the provided `resourceVersion`\"\n  and the bookmark event is send when the state is synced\n  to a `resourceVersion` at least as fresh as the one provided by the ListOptions.\n  If `resourceVersion` is unset, this is interpreted as \"consistent read\" and the\n  bookmark event is send when the state is synced at least to the moment\n  when request started being processed.\n- `resourceVersionMatch` set to any other value or unset\n  Invalid error is returned.\n\nDefaults to true if `resourceVersion=\"\"` or `resourceVersion=\"0\"` (for backward compatibility reasons) and to false otherwise.",
+          "in": "query",
+          "name": "sendInitialEvents",
+          "type": "boolean",
+          "uniqueItems": true
+        },
         {
           "description": "Timeout for the list/watch call. This limits the duration of the call, regardless of any activity or inactivity.",
           "in": "query",
@@ -59153,6 +59174,13 @@
           "type": "string",
           "uniqueItems": true
         },
+        {
+          "description": "`sendInitialEvents=true` may be set together with `watch=true`. In that case, the watch stream will begin with synthetic events to produce the current state of objects in the collection. Once all such events have been sent, a synthetic \"Bookmark\" event  will be sent. The bookmark will report the ResourceVersion (RV) corresponding to the set of objects, and be marked with `\"k8s.io/initial-events-end\": \"true\"` annotation. Afterwards, the watch stream will proceed as usual, sending watch events corresponding to changes (subsequent to the RV) to objects watched.\n\nWhen `sendInitialEvents` option is set, we require `resourceVersionMatch` option to also be set. The semantic of the watch request is as following: - `resourceVersionMatch` = NotOlderThan\n  is interpreted as \"data at least as new as the provided `resourceVersion`\"\n  and the bookmark event is send when the state is synced\n  to a `resourceVersion` at least as fresh as the one provided by the ListOptions.\n  If `resourceVersion` is unset, this is interpreted as \"consistent read\" and the\n  bookmark event is send when the state is synced at least to the moment\n  when request started being processed.\n- `resourceVersionMatch` set to any other value or unset\n  Invalid error is returned.\n\nDefaults to true if `resourceVersion=\"\"` or `resourceVersion=\"0\"` (for backward compatibility reasons) and to false otherwise.",
+          "in": "query",
+          "name": "sendInitialEvents",
+          "type": "boolean",
+          "uniqueItems": true
+        },
         {
           "description": "Timeout for the list/watch call. This limits the duration of the call, regardless of any activity or inactivity.",
           "in": "query",
diff --git a/api/openapi-spec/v3/apis__certificates.k8s.io__v1alpha1_openapi.json b/api/openapi-spec/v3/apis__certificates.k8s.io__v1alpha1_openapi.json
index 02e9c53b0f4..cbfea384115 100644
--- a/api/openapi-spec/v3/apis__certificates.k8s.io__v1alpha1_openapi.json
+++ b/api/openapi-spec/v3/apis__certificates.k8s.io__v1alpha1_openapi.json
@@ -629,7 +629,7 @@
               "default": "",
               "type": "string"
             },
-            "description": "Annotations is an unstructured key value map stored with a resource that may be set by external tools to store and retrieve arbitrary metadata. They are not queryable and should be preserved when modifying objects. More info: http://kubernetes.io/docs/user-guide/annotations",
+            "description": "Annotations is an unstructured key value map stored with a resource that may be set by external tools to store and retrieve arbitrary metadata. They are not queryable and should be preserved when modifying objects. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations",
             "type": "object"
           },
           "creationTimestamp": {
@@ -677,7 +677,7 @@
               "default": "",
               "type": "string"
             },
-            "description": "Map of string keys and values that can be used to organize and categorize (scope and select) objects. May match selectors of replication controllers and services. More info: http://kubernetes.io/docs/user-guide/labels",
+            "description": "Map of string keys and values that can be used to organize and categorize (scope and select) objects. May match selectors of replication controllers and services. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels",
             "type": "object"
           },
           "managedFields": {
@@ -693,11 +693,11 @@
             "type": "array"
           },
           "name": {
-            "description": "Name must be unique within a namespace. Is required when creating resources, although some resources may allow a client to request the generation of an appropriate name automatically. Name is primarily intended for creation idempotence and configuration definition. Cannot be updated. More info: http://kubernetes.io/docs/user-guide/identifiers#names",
+            "description": "Name must be unique within a namespace. Is required when creating resources, although some resources may allow a client to request the generation of an appropriate name automatically. Name is primarily intended for creation idempotence and configuration definition. Cannot be updated. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names#names",
             "type": "string"
           },
           "namespace": {
-            "description": "Namespace defines the space within which each name must be unique. An empty namespace is equivalent to the \"default\" namespace, but \"default\" is the canonical representation. Not all objects are required to be scoped to a namespace - the value of this field for those objects will be empty.\n\nMust be a DNS_LABEL. Cannot be updated. More info: http://kubernetes.io/docs/user-guide/namespaces",
+            "description": "Namespace defines the space within which each name must be unique. An empty namespace is equivalent to the \"default\" namespace, but \"default\" is the canonical representation. Not all objects are required to be scoped to a namespace - the value of this field for those objects will be empty.\n\nMust be a DNS_LABEL. Cannot be updated. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces",
             "type": "string"
           },
           "ownerReferences": {
@@ -723,7 +723,7 @@
             "type": "string"
           },
           "uid": {
-            "description": "UID is the unique in time and space value for this object. It is typically generated by the server on successful creation of a resource and is not allowed to change on PUT operations.\n\nPopulated by the system. Read-only. More info: http://kubernetes.io/docs/user-guide/identifiers#uids",
+            "description": "UID is the unique in time and space value for this object. It is typically generated by the server on successful creation of a resource and is not allowed to change on PUT operations.\n\nPopulated by the system. Read-only. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names#uids",
             "type": "string"
           }
         },
@@ -752,12 +752,12 @@
           },
           "name": {
             "default": "",
-            "description": "Name of the referent. More info: http://kubernetes.io/docs/user-guide/identifiers#names",
+            "description": "Name of the referent. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names#names",
             "type": "string"
           },
           "uid": {
             "default": "",
-            "description": "UID of the referent. More info: http://kubernetes.io/docs/user-guide/identifiers#uids",
+            "description": "UID of the referent. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names#uids",
             "type": "string"
           }
         },
@@ -899,7 +899,7 @@
             "type": "integer"
           },
           "uid": {
-            "description": "UID of the resource. (when there is a single resource which can be described). More info: http://kubernetes.io/docs/user-guide/identifiers#uids",
+            "description": "UID of the resource. (when there is a single resource which can be described). More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names#uids",
             "type": "string"
           }
         },
@@ -1374,6 +1374,15 @@
               "uniqueItems": true
             }
           },
+          {
+            "description": "`sendInitialEvents=true` may be set together with `watch=true`. In that case, the watch stream will begin with synthetic events to produce the current state of objects in the collection. Once all such events have been sent, a synthetic \"Bookmark\" event  will be sent. The bookmark will report the ResourceVersion (RV) corresponding to the set of objects, and be marked with `\"k8s.io/initial-events-end\": \"true\"` annotation. Afterwards, the watch stream will proceed as usual, sending watch events corresponding to changes (subsequent to the RV) to objects watched.\n\nWhen `sendInitialEvents` option is set, we require `resourceVersionMatch` option to also be set. The semantic of the watch request is as following: - `resourceVersionMatch` = NotOlderThan\n  is interpreted as \"data at least as new as the provided `resourceVersion`\"\n  and the bookmark event is send when the state is synced\n  to a `resourceVersion` at least as fresh as the one provided by the ListOptions.\n  If `resourceVersion` is unset, this is interpreted as \"consistent read\" and the\n  bookmark event is send when the state is synced at least to the moment\n  when request started being processed.\n- `resourceVersionMatch` set to any other value or unset\n  Invalid error is returned.\n\nDefaults to true if `resourceVersion=\"\"` or `resourceVersion=\"0\"` (for backward compatibility reasons) and to false otherwise.",
+            "in": "query",
+            "name": "sendInitialEvents",
+            "schema": {
+              "type": "boolean",
+              "uniqueItems": true
+            }
+          },
           {
             "description": "Timeout for the list/watch call. This limits the duration of the call, regardless of any activity or inactivity.",
             "in": "query",
@@ -1495,6 +1504,15 @@
               "uniqueItems": true
             }
           },
+          {
+            "description": "`sendInitialEvents=true` may be set together with `watch=true`. In that case, the watch stream will begin with synthetic events to produce the current state of objects in the collection. Once all such events have been sent, a synthetic \"Bookmark\" event  will be sent. The bookmark will report the ResourceVersion (RV) corresponding to the set of objects, and be marked with `\"k8s.io/initial-events-end\": \"true\"` annotation. Afterwards, the watch stream will proceed as usual, sending watch events corresponding to changes (subsequent to the RV) to objects watched.\n\nWhen `sendInitialEvents` option is set, we require `resourceVersionMatch` option to also be set. The semantic of the watch request is as following: - `resourceVersionMatch` = NotOlderThan\n  is interpreted as \"data at least as new as the provided `resourceVersion`\"\n  and the bookmark event is send when the state is synced\n  to a `resourceVersion` at least as fresh as the one provided by the ListOptions.\n  If `resourceVersion` is unset, this is interpreted as \"consistent read\" and the\n  bookmark event is send when the state is synced at least to the moment\n  when request started being processed.\n- `resourceVersionMatch` set to any other value or unset\n  Invalid error is returned.\n\nDefaults to true if `resourceVersion=\"\"` or `resourceVersion=\"0\"` (for backward compatibility reasons) and to false otherwise.",
+            "in": "query",
+            "name": "sendInitialEvents",
+            "schema": {
+              "type": "boolean",
+              "uniqueItems": true
+            }
+          },
           {
             "description": "Timeout for the list/watch call. This limits the duration of the call, regardless of any activity or inactivity.",
             "in": "query",
@@ -2191,6 +2209,15 @@
             "uniqueItems": true
           }
         },
+        {
+          "description": "`sendInitialEvents=true` may be set together with `watch=true`. In that case, the watch stream will begin with synthetic events to produce the current state of objects in the collection. Once all such events have been sent, a synthetic \"Bookmark\" event  will be sent. The bookmark will report the ResourceVersion (RV) corresponding to the set of objects, and be marked with `\"k8s.io/initial-events-end\": \"true\"` annotation. Afterwards, the watch stream will proceed as usual, sending watch events corresponding to changes (subsequent to the RV) to objects watched.\n\nWhen `sendInitialEvents` option is set, we require `resourceVersionMatch` option to also be set. The semantic of the watch request is as following: - `resourceVersionMatch` = NotOlderThan\n  is interpreted as \"data at least as new as the provided `resourceVersion`\"\n  and the bookmark event is send when the state is synced\n  to a `resourceVersion` at least as fresh as the one provided by the ListOptions.\n  If `resourceVersion` is unset, this is interpreted as \"consistent read\" and the\n  bookmark event is send when the state is synced at least to the moment\n  when request started being processed.\n- `resourceVersionMatch` set to any other value or unset\n  Invalid error is returned.\n\nDefaults to true if `resourceVersion=\"\"` or `resourceVersion=\"0\"` (for backward compatibility reasons) and to false otherwise.",
+          "in": "query",
+          "name": "sendInitialEvents",
+          "schema": {
+            "type": "boolean",
+            "uniqueItems": true
+          }
+        },
         {
           "description": "Timeout for the list/watch call. This limits the duration of the call, regardless of any activity or inactivity.",
           "in": "query",
@@ -2343,6 +2370,15 @@
             "uniqueItems": true
           }
         },
+        {
+          "description": "`sendInitialEvents=true` may be set together with `watch=true`. In that case, the watch stream will begin with synthetic events to produce the current state of objects in the collection. Once all such events have been sent, a synthetic \"Bookmark\" event  will be sent. The bookmark will report the ResourceVersion (RV) corresponding to the set of objects, and be marked with `\"k8s.io/initial-events-end\": \"true\"` annotation. Afterwards, the watch stream will proceed as usual, sending watch events corresponding to changes (subsequent to the RV) to objects watched.\n\nWhen `sendInitialEvents` option is set, we require `resourceVersionMatch` option to also be set. The semantic of the watch request is as following: - `resourceVersionMatch` = NotOlderThan\n  is interpreted as \"data at least as new as the provided `resourceVersion`\"\n  and the bookmark event is send when the state is synced\n  to a `resourceVersion` at least as fresh as the one provided by the ListOptions.\n  If `resourceVersion` is unset, this is interpreted as \"consistent read\" and the\n  bookmark event is send when the state is synced at least to the moment\n  when request started being processed.\n- `resourceVersionMatch` set to any other value or unset\n  Invalid error is returned.\n\nDefaults to true if `resourceVersion=\"\"` or `resourceVersion=\"0\"` (for backward compatibility reasons) and to false otherwise.",
+          "in": "query",
+          "name": "sendInitialEvents",
+          "schema": {
+            "type": "boolean",
+            "uniqueItems": true
+          }
+        },
         {
           "description": "Timeout for the list/watch call. This limits the duration of the call, regardless of any activity or inactivity.",
           "in": "query",

```

pacoxu commented:
/hold
To avoid ci-bot retesting automatically. (Feel free to unhold once the CI is happy).

marosset commented:
/test pull-kubernetes-verify

marosset commented:
/hold - https://github.com/kubernetes/enhancements/issues/3257#issuecomment-1469030391

ahmedtd commented:
Rebased

pacoxu commented:
/lgtm

ahmedtd commented:
Rebased again

ahmedtd commented:
/test pull-kubernetes-unit

/test pull-kubernetes-e2e-gce-cos-alpha-features

liggitt commented:
/lgtm

salaxander commented:
/unhold
/milestone v1.27

sftim commented:
Changelog suggestion
```diff
-The new certificates.k8s.io/v1alpha1 ClusterTrustBundle type may be used to distribute X.509 trust anchors to workloads within the cluster.
+Added a new alpha API: ClusterTrustBundle (`certificates.k8s.io/v1alpha1`).
+A ClusterTrustBundle may be used to distribute [X.509](https://www.itu.int/rec/T-REC-X.509) trust anchors to workloads within the cluster.
```

liggitt commented:
xref exception approval in https://groups.google.com/g/kubernetes-sig-release/c/kkA4F4LvMrM/m/ojyaDWowBAAJ

ahmedtd commented:
Applied changelog suggestion.
